### PR TITLE
Extend markdown toolbar

### PR DESCRIPTION
**Pull request by alex-ju**
This PR:
- moves markdown-toolbar in a component
- groups toolbar items in sections using the `details` component and add links to the relevant guidance (the content of these sections will be reviewed with a content designer and potentially read from a separate file)
- ~prevents markdown contextual guidance to overlap footer or other components on small screens~
- extends markdown-toolbar with the following snippets:
  - heading 2 (`##`), heading 3 (`###`), heading 4 (`####`)
  - ~call to action~ (`$CTA [Start](url) %CTA`)
  - link email (`<john.doe@mail.tld>`)
  - ~address~ (`$A 10 Whitechapel $A`)
- ~configures marked renderer to~
  - support line breaks
  - render tables
  - sanitise user input
- removes the absolute position of contextual guidance

[Trello card](https://trello.com/c/yKPf7Kd6)

### CODE REVIEW

File: app/views/components/_markdown-toolbar.html.erb [hunk 1]
```diff
@@ -0,0 +1,71 @@
+<markdown-toolbar class="app-c-markdown-toolbar" for="<%= html_for %>">
+  <%= render "govuk_publishing_components/components/details", {
+      title: "Headings and sub-headings"
```
Comment (benthorner) [inline app/views/components/_markdown-toolbar.html.erb]: Which ones are sub-headings?

File: app/views/components/_markdown-toolbar.html.erb [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+  <%= render "govuk_publishing_components/components/details", {
+      title: "Blockquotes"
+  } do %>
+  <ul class="app-c-markdown-toolbar__list">
+    <li class="app-c-markdown-toolbar__item">
+      <md-quote class="app-c-markdown-toolbar__button">Blockquotes</md-quote>
+    </li>
+  </ul>
+  <a href="https://www.gov.uk/guidance/how-to-publish-on-gov-uk/markdown#blockquotes" class="govuk-link" target="_blank"> Guidance on blockquotes</a>
```
Comment (benthorner) [inline app/views/components/_markdown-toolbar.html.erb]: Interestingly this one seems to toggle them on or off, whereas the other tool buttons just toggle-on. Is that expected?

File: app/views/components/_markdown-toolbar.html.erb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  <% end %>
+
+  <%= render "govuk_publishing_components/components/details", {
```
Comment (benthorner) [inline app/views/components/_markdown-toolbar.html.erb]: I feel like we shouldn't add this feature until it works end-to-end - if it's in the toolbar then I'd expect it to work in preview mode, otherwise it's liable to make the user feel like they did something wrong. Can we leave it out of this PR and add it separately?

File: app/views/components/_markdown-toolbar.html.erb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  } do %>
+  <ul class="app-c-markdown-toolbar__list">
+    <li class="app-c-markdown-toolbar__item">
+      <md-link class="app-c-markdown-toolbar__button">Link</md-link>
+    </li>
+    <li class="app-c-markdown-toolbar__item">
+      <md-link-email class="app-c-markdown-toolbar__button">Email link</md-link-email>
+    </li>
+    <li class="app-c-markdown-toolbar__item">
+      <md-link-cta class="app-c-markdown-toolbar__button">Call to action</md-link-cta>
```
Comment (benthorner) [inline app/views/components/_markdown-toolbar.html.erb]: I feel like we shouldn't add this feature until it works end-to-end - if it's in the toolbar then I'd expect it to work in preview mode, otherwise it's liable to make the user feel like they did something wrong. Can we leave it out of this PR and add it separately?

File: app/assets/stylesheets/components/_contextual-guidance.scss [hunk 5]
```diff
@@ -1,10 +1,18 @@
...
+  @include govuk-responsive-margin(6, "bottom");
+
+  @include mq ($from: desktop) {
+    padding-top: govuk-spacing(6);
+  }
 }
 
 .app-c-contextual-guidance {
-  position: absolute;
+  @include mq ($from: desktop) {
```
Comment (benthorner) [inline app/assets/stylesheets/components/_contextual-guidance.scss]: Could you explain briefly what the change here is doing?

File: app/assets/javascripts/components/markdown-editor.js [hunk 6]
```diff
@@ -68,6 +81,16 @@ MarkdownEditor.prototype.toggle = function (element) {
   }
 }
 
+MarkdownEditor.prototype.setContextualGuidancePosition = function () {
+  var guidanceId = this.$input.getAttribute('data-contextual-guidance')
+  if (guidanceId) {
+    var $guidance = document.querySelector('#' + guidanceId + ' .app-c-contextual-guidance')
```
Comment (benthorner) [inline app/assets/javascripts/components/markdown-editor.js]: How come this is overriding the style 'position: absolute' - could we not just change the style?

File: app/views/documents/fields/_govspeak_input.html.erb [hunk 7]
```diff
@@ -11,7 +11,7 @@
         id: "govspeak-editor",
         name: "document[contents][#{schema.id}]",
         value: document.contents[schema.id],
-        rows: 30,
+        rows: 25,
```
Comment (benthorner) [inline app/views/documents/fields/_govspeak_input.html.erb]: Why has this changed?

File: app/assets/javascripts/components/markdown-editor.js [hunk 8]
```diff
@@ -33,16 +36,26 @@ MarkdownEditor.prototype.handlePreviewButton = function (event) {
...
-      $preview.innerHTML = content
-    }
-  })
-
+  if (text) {
+    // Render markdown
+    window.marked(
+      text,
+      {
+        gfm: true,
```
Comment (benthorner) [inline app/assets/javascripts/components/markdown-editor.js]: I'm not really sure what this one's doing. I tried 'hello<newline>there' and 'hello<space><space><newline>there' and in the latter the editor inserted a '<dot>' in edit mode. Perhaps this bit (together with these other options), would be better in a separate PR which gives more focus to the features they enable (and maybe adds it to the toolbar?).

File: app/views/components/_markdown-toolbar.html.erb (hunk 1 above)
Comment (alex-ju) [reply app/views/components/_markdown-toolbar.html.erb]: This is the current copy. I will update with the rest of the guidance.

File: app/views/components/_markdown-toolbar.html.erb (hunk 3 above)
Comment (alex-ju) [reply app/views/components/_markdown-toolbar.html.erb]: Sure, we can leave them aside for now.

File: app/assets/stylesheets/components/_contextual-guidance.scss (hunk 5 above)
Comment (alex-ju) [reply app/assets/stylesheets/components/_contextual-guidance.scss]: On mobile, the guidance shouldn't be positioned absolute, as they will overlap with the next element, so switched to position them absolute only on desktop when we don't want the height of the guidance to push the next components down the page.

File: app/assets/javascripts/components/markdown-editor.js (hunk 6 above)
Comment (alex-ju) [reply app/assets/javascripts/components/markdown-editor.js]: We're overwriting the position only for the contextual guidance that relates to the markdown editor as it contains the expandable toolbar which can overlap the footer.

File: app/views/documents/fields/_govspeak_input.html.erb (hunk 7 above)
Comment (alex-ju) [reply app/views/documents/fields/_govspeak_input.html.erb]: There isn't a fixed value for this. I initially set this arbitrary at 30, but it seems to be to tall. Will discuss with the designer to figure out what will be a 'good' height for the editor.

File: app/assets/javascripts/components/markdown-editor.js (hunk 8 above)
Comment (alex-ju) [reply app/assets/javascripts/components/markdown-editor.js]: Yes, I can remove them from this PR and add them when we decide on what we want to show in the markdown preview.

File: app/views/components/_markdown-toolbar.html.erb (hunk 2 above)
Comment (alex-ju) [reply app/views/components/_markdown-toolbar.html.erb]: Not sure I get this one. The 'Blockquotes' button should behave the same with the rest.
